File: design_patterns/C++/Behavioral/state.cxx
#include <iostream>
#include <string>
#include <algorithm>

class WritingState {
  public:
	virtual void write(std::string words) = 0;
};

class UpperCase: public WritingState {
  public:
	void write(std::string words) {
		std::transform(words.begin(), words.end(), words.begin(), ::toupper);
		std::cout << words << std::endl;
	}
};

class LowerCase: public WritingState {
  public:
	void write(std::string words) {
		std::transform(words.begin(), words.end(), words.begin(), ::tolower);
		std::cout << words << std::endl;
	}
};

class Default: public WritingState {
  public:
	void write(std::string words) {
		std::cout << words << std::endl;
	}
};

class TextEditor {
  protected:
	WritingState* state = nullptr;
  public:
	TextEditor(WritingState* state) {
		this->state = state;
	}

	void setState(WritingState* state) {
		this->state = state;
	}

	void type(std::string words) {
		this->state->write(words);
	}
};

int main() {
	auto editor = new TextEditor(new Default());
	editor->type("First line");
	editor->setState(new UpperCase());
	editor->type("Second line");
	editor->type("Third line");
	editor->setState(new LowerCase());
	editor->type("Fourth line");
	editor->type("Fifth line");
	// ===========
	// First line
	// SECOND LINE
	// THIRD LINE
	// fourth line
	// fifth line
	// ===========
	return 0;
}

File: design_patterns/Python/Behavioral/state.py
class WritingState:

    def write(self, words):
        raise NotImplementedError


class UpperCase(WritingState):

    def write(self, words):
        print(words.upper())


class LowerCase(WritingState):

    def write(self, words):
        print(words.lower())


class Default(WritingState):

    def write(self, words):
        print(words)


class TextEditor:

    def __init__(self, state):
        self.state = state

    def set_state(self, state):
        self.state = state

    def type(self, words):
        self.state.write(words)


if __name__ == '__main__':

    editor = TextEditor(Default())
    editor.type('First line')

    editor.set_state(UpperCase())
    editor.type('Second line')
    editor.type('Third line')

    editor.set_state(LowerCase())
    editor.type('Fourth line')
    editor.type('Fifth line')

    # First line
    # SECOND LINE
    # THIRD LINE
    # fourth line
    # fifth line
